scene/main/scene_tree.h

```diff
@@ -437,6 +447,9 @@ class SceneTree : public MainLoop {
 	void client_physics_interpolation_add_spatial(SelfList<Spatial> *p_elem);
 	void client_physics_interpolation_remove_spatial(SelfList<Spatial> *p_elem);
 
+	void notify_canvas_parent_children_moved(Node &p_parent, uint32_t p_first_child, uint32_t p_last_child_plus_one);
+	void notify_canvas_parent_child_count_reduced(Node &p_parent);
```

Comment:
Can you elaborate? I'm not sure I understand reasoning, maybe I'm missing something (then again it's a while since I wrote this PR :grin: ).

Reply:
You're passing `Node &` rather than `Node *`, haven't seen that anywhere else

Reply:
https://stackoverflow.com/questions/7058339/when-should-i-use-pointers-instead-of-references-in-api-design

https://isocpp.org/wiki/faq/references#:~:text=Use%20references%20when%20you%20can,and%20pointers%20on%20the%20inside.

https://www.geeksforgeeks.org/pointers-vs-references-cpp/

> Use references: 
> In function parameters and return types.
> Use pointers: 
> * If pointer arithmetic or passing a NULL pointer is needed. For example, for arrays (Note that accessing an array is implemented using pointer arithmetic).
> * To implement data structures like a linked list, a tree, etc. and their algorithms. This is so because, in order to point to different cells, we have to use the concept of pointers.

It is usually regarded as good practice in c++ to use references in preference to pointers _wherever possible_. A reference is treated by the compiler as essentially a pointer that CANNOT be NULL. It may also lead to more optimization opportunities in some situations.

This both removes the need for `ERR_FAIL` checks within the function, and prevents the program compiling in a lot of situations that are likely to cause a NULL to be passed, which excludes an entire class of bugs (as you have to either pass an object, or explicitly dereference a pointer on calling, and `*this` is guaranteed to be valid on non-static function as used here).

The reason (as I understand it) Godot uses pointers more often than you would see in most codebases is largely due to _binding_ (for other languages), and the macros which often support pointers but not non-const references. In internal functions that are not bound, reference is preferable imo.

Reply:
Oh I'm aware of the benefits 🙂 but it's unorthodox and we do treat our pointers *somewhat* differently than plain ones, but should be safe here

Reply:
On the other hand, what we do do normally do is prefix with `r_` for the non-const references.

I have a feeling I didn't change the prefix here because it was just to resolve a compile error with something internally needing to be non-const, I'll double check this.

Reply:
`notify_canvas_parent_children_moved()` does need non-const reference, due to calling `set_canvas_parent_id()`. I've changed parameter name to `r_parent` for consistency.

`notify_canvas_parent_child_count_reduced()` turned out to only need a const reference so I've fixed this.